### ISSUE TITLE
[no-unnecessary-type-assertion] false positive for nullable generic

### ISSUE THREAD

**Issue by mmkal** [typescript-eslint/typescript-eslint#2727]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": ["error"]
  }
}
```

```typescript
const getLengthOrThrow = <T extends { length: number }>(val?: T) => val!.length;
```

```json
{
	"include": [
		"src"
	],
	"compilerOptions": {
		"outDir": "lib",
		"target": "es2018",
		"module": "commonjs",
		"moduleResolution": "node",
		"resolveJsonModule": true,
		"declaration": true,
		"sourceMap": true,
		"pretty": true,
		"newLine": "lf",
		"strictNullChecks": true,
		"stripInternal": true,
		"noImplicitReturns": true,
		"noUnusedLocals": false,
		"noUnusedParameters": false,
		"noFallthroughCasesInSwitch": true,
		"noEmitOnError": true,
		"skipLibCheck": true
	}
}
```

(only `strictNullChecks` is important here, I believe)

**Expected Result**

No error.

**Actual Result**

`This assertion is unnecessary since it does not change the type of the expression. eslint @typescript-eslint/no-unnecessary-type-assertion`

**Additional Info**

No.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.6.0` |
| `@typescript-eslint/parser`        | `4.6.0` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `7.12.0` |
| `node`                             | `12.18.0` |

**Comment by bradzacher** [comment]
I am unable to repro this against master.

**Comment by mmkal** [comment]
@bradzacher this didn't repro on CI. I'm still seeing it locally even after restarting VSCode but I think this it's still just a caching issue. I recently turned on `strictNullChecks` on the project, so it might be the `--cache` flag making the old error still show up. Closing.
